File: examples/05-registry/after.ts
/**
 * Example 05 — Factory Method / Strategy, with the pattern.
 *
 * Registration lives next to each implementation, so a new rail is a new file and
 * one `register` call. The switch never gets reopened, and an unknown key throws
 * with the key in the message.
 *
 * Run it:  node examples/05-registry/after.ts
 */
import assert from 'node:assert/strict';
import { registry } from '../../packages/ts/src/index.ts';

type Rail = { send(cents: number): string };

const rails = registry<{
  stripe: () => Rail;
  spei: (clabe: string) => Rail;
  paypal: () => Rail;
}>();

// Each of these lines would live in the file that implements that rail.
rails.register('stripe', () => ({ send: (cents) => `stripe:${cents}` }));
rails.register('spei', (clabe) => ({ send: (cents) => `spei:${clabe}:${cents}` }));

const results = [rails.create('stripe').send(1999), rails.create('spei', '0123').send(1999)];
results.forEach((result) => console.log(`  ${result}`));

let refused = '';
try {
  rails.create('paypal');
} catch (error) {
  refused = (error as Error).message;
}
console.log(`  desconocido → ${refused}`);

assert.deepEqual(results, ['stripe:1999', 'spei:0123:1999']);
assert.match(refused, /nothing registered for "paypal"/);

// The rail that arrived later: a new file, one line, nothing else edited.
rails.register('paypal', () => ({ send: (cents) => `paypal:${cents}` }));
assert.equal(rails.create('paypal').send(1999), 'paypal:1999');
assert.deepEqual(rails.keys().sort(), ['paypal', 'spei', 'stripe']);

console.log('\n✅ un riel nuevo es un archivo nuevo, y lo desconocido truena con la clave');

File: examples/05-registry/after.py
"""Video 05 — Factory Method / Strategy, with the pattern.

Registration lives next to each implementation — and in Python it can be a
decorator, so the registration sits directly above the thing it registers.

Run it:  PYTHONPATH=packages/py python3 examples/05-registry/after.py
"""

from gof_patterns import Registry

rails: Registry = Registry()


# The decorator form: the registration is right above the implementation, which is
# what makes "a new rail is a new file" true rather than aspirational.
@rails.register("stripe")
def stripe():
    return lambda cents: f"stripe:{cents}"


@rails.register("spei")
def spei(clabe: str):
    return lambda cents: f"spei:{clabe}:{cents}"


results = [rails.create("stripe")(1999), rails.create("spei", "0123")(1999)]
for result in results:
    print(f"  {result}")

refused = ""
try:
    rails.create("paypal")
except KeyError as error:
    refused = str(error)
print(f"  desconocido → {refused}")

assert results == ["stripe:1999", "spei:0123:1999"]
assert "paypal" in refused

# Still a normal function, so it is testable on its own.
assert stripe()(1) == "stripe:1"
assert sorted(rails.keys()) == ["spei", "stripe"]

print("\n✅ el registro vive junto a la implementación, y lo desconocido truena")
